### PR TITLE
Faheel/test enhancements

### PR DESCRIPTION
**Pull request by faheelsattar**
## 📝 Summary

This PR introduces several critical test cases that improve the current  MEV-Boost's testing. The new tests focus on relay status, timeout handling, multiple relay scenarios, and payload validations associated with their respective handlers, ensuring more reliable operation and taking care of edge cases which were previously uncovered. 

## ⛱ Motivation and Context

Before starting, I considered two potential approaches:
1. Introduce full e2e tests  and integrate them into CI using the builder-playground.
2. Improve coverage and address untested scenarios within the existing test suite.

I started with the first approach and spent sometime integrating and writing full e2e tests via builder-playground. Similar to what we have for rbuilder. I successfully got all the things working. However, I encountered a blocker where rbuilder was unable to send out bids. I briefly started to look for ways to get it fixed and have it working but was bounded due to time constraints. An integration with Kurtosis was also an option but it was discouraged in the task description due to time constraints. 

I then opted for option 2 which was much simpler and had some areas of improvement. Coverage tool helped to uncover some areas of improvement, but also manually looking at the system from a high level and how different components in PBS pipeline were interacting further helped me improve the current testing suite.

### Key Test Cases Added:
1. **Relay check disabled:**
      This tests sets the ```relayCheck``` to false and then verifies how ```handleStatus``` handler responds to the status.
2. **Relay timeout:**
      This test scenario tests how does ```handleStatus``` handler responds to the requests with varying response time from the relays.
3. **Multiple relays with different timeouts:** 
     This test essentially checks how ```handleStatus``` handles the requests incase of multiple relays with varying response time. What the response is and how it gets handled?
4. **Use header with highest value with varying relay timeouts:**
     This test ensures the selection of highest value bid regardless of response time from the configured relays. Verifies proper handling of concurrent relay responses.
5. **Returns successful response when one relay fails after max retries reached and another responds:**
    This test essentially makes sure that the in case of multiple relays, if one of the relays fails to respond with a payload even after trying maximum amount of times, the payload is still delivered from the other responsive/active relays.
6. **Requesting a payload without getting header:**
     Ensures that validators cannot request payloads without first requesting headers for the corresponding slot and block hash
7. **Requesting payload for incorrect slot and block hash:**
   Ensures that validators aren't allowed to receive payload for incorrect slot and block hashes. They may have requested the headers with different slot and block hashes, but when requesting the payloads provided incorrect values causing in a cache miss and in turn not receiving the payload. 

There are some refactors which can be added to the current test-suite but I think it will be much better to introduce it as a seperate PR.

## 📚 References

[mev-boost docs](https://docs.flashbots.net/flashbots-mev-boost/introduction)
[builder-api spec](https://ethereum.github.io/builder-specs/#/Builder)
[mev-boost-relay specs](https://flashbots.github.io/relay-specs/)

---

## ✅ I have run these commands

* [x] `make lint`
* [x] `make test-race`
* [x] `go mod tidy`

### CODE REVIEW

File: server/service_test.go [hunk 1]
```diff
@@ -216,6 +216,48 @@ func TestStatus(t *testing.T) {
...
+		backend.relays[1].ResponseDelay = 300 * time.Millisecond // Slow
+		backend.relays[2].ResponseDelay = 150 * time.Millisecond // Medium
+
+		header := make(http.Header)
+		header.Set(HeaderAccept, MediaTypeJSON)
+		path := "/eth/v1/builder/status"
+
+		rr := backend.request(t, http.MethodGet, path, header, nil)
+		require.Equal(t, http.StatusOK, rr.Code)
+	})
```
Comment (jtraglia) [inline server/service_test.go:260]: Would it be possible to check that there were 2 responses? I don't recall if we have that much introspection. It's okay if the answer to this is no.

File: server/service_test.go [hunk 2]
```diff
@@ -1082,6 +1184,94 @@ func TestGetPayload(t *testing.T) {
...
 		require.Equal(t, http.StatusBadGateway, rr.Code, rr.Body.String())
 	})
+
+	t.Run("Returns successful response when one relay fails after max retries reached and another responds", func(t *testing.T) {
+		header := make(http.Header)
+		header.Set(HeaderAccept, MediaTypeJSON)
+
+		backend := newTestBackend(t, 2, 2*time.Second)
+
+		// Add the bid to the service
```
Comment (jtraglia) [inline server/service_test.go]: Just a nit, I've noticed that some comments have punctuation and some do not.

File: server/service_test.go [hunk 3]
```diff
@@ -755,6 +797,66 @@ func TestGetHeaderBids(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, uint256.NewInt(12345), value)
 	})
+
+	t.Run("Use header with highest value with varying relay timeouts", func(t *testing.T) {
+		header := make(http.Header)
+		header.Set(HeaderAccept, MediaTypeJSON)
+
+		// Create backend and register 3 relays.
+		backend := newTestBackend(t, 3, time.Second)
```
Comment (jtraglia) [inline server/service_test.go:806]: I would be interested in a test where the highest value bid arrives too late, after the `time.Second` curfew.

File: server/service_test.go [hunk 4]
```diff
@@ -1082,6 +1184,94 @@ func TestGetPayload(t *testing.T) {
...
+		require.Equal(t, payload.Message.Body.ExecutionPayloadHeader.BlockHash, resp.Deneb.ExecutionPayload.BlockHash)
+	})
+
+	t.Run("Requesting payload without getting header", func(t *testing.T) {
+		header := make(http.Header)
+		header.Set(HeaderAccept, MediaTypeJSON)
+
+		backend := newTestBackend(t, 1, time.Second)
+		rr := backend.request(t, http.MethodPost, path, header, payload)
+		// no bids added to the bid cache since GetHeader request wasnt made resulting in empty bids cache
```
Comment (jtraglia) [inline server/service_test.go]: For consistency, this comment should start with an uppercase letter.

File: server/service_test.go [hunk 5]
```diff
@@ -216,6 +216,48 @@ func TestStatus(t *testing.T) {
 		require.NotEmpty(t, rr.Header().Get("X-MEVBoost-Version"))
 		require.Equal(t, 0, backend.relays[0].GetRequestCount(path))
 	})
+
+	t.Run("Relay check disabled", func(t *testing.T) {
+		backend := newTestBackend(t, 1, time.Second)
+		// since newTestBackend sets relayCheck to true for all requests,
+		// we need to override it to check the status when it is false
```
Comment (jtraglia) [inline server/service_test.go]: Same there here, regarding starting with an uppercase letter.

**Review by jtraglia — COMMENTED** [review]
Hey @faheelsattar, these tests look valuable, nice, and clean 🙂 I appreciate the clear comments a lot, they make understanding what's going on a lot easier. I don't see anything wrong with your tests, my review comments are just for small nits.

File: server/service_test.go (hunk 2 above)
Comment (faheelsattar) [reply server/service_test.go]: Thanks for the suggestion, have fixed for all of them https://github.com/faheelsattar/mev-boost/pull/1/commits/a3213bb7c894ca78ed24bad2c020af58b28396c1

File: server/service_test.go (hunk 1 above)
Comment (faheelsattar) [reply server/service_test.go:260]: You are right, we don't have that level of introspection. While we can verify that requests were sent to the relays using ```GetRequestCount```, we cannot verify if responses were actually received from each relay currently.

File: server/service_test.go (hunk 3 above)
Comment (faheelsattar) [reply server/service_test.go:806]: Great suggestion! Added the case https://github.com/faheelsattar/mev-boost/pull/1/commits/c7c1f78bc22ff17d2a91c27d58e52a314f94189d
